tee-worker/core/parentchain/indirect-calls-executor/src/litentry/mod.rs

```diff
@@ -0,0 +1,160 @@
+use codec::{Decode, Encode, Error as CodecError};
+use itp_node_api::{
+	api_client::{
+		ParentchainUncheckedExtrinsic, PlainTip, SubstrateDefaultSignedExtra, UncheckedExtrinsicV4,
+	},
+	metadata::{
+		pallet_imp::IMPCallIndexes, pallet_teerex::TeerexCallIndexes, pallet_vcmp::VCMPCallIndexes,
+		provider::AccessNodeMetadata, Error as MetadataError,
+	},
+};
+use itp_sgx_crypto::{key_repository::AccessKey, ShieldingCryptoDecrypt, ShieldingCryptoEncrypt};
+use itp_stf_executor::traits::StfEnclaveSigning;
+use itp_top_pool_author::traits::AuthorApi;
+use itp_types::H256;
+
+pub mod create_identity;
+pub mod remove_identity;
+pub mod request_vc;
+pub mod set_user_shielding_key;
+pub mod verify_identity;
+
+pub(crate) trait Executor<
+	ShieldingKeyRepository,
+	StfEnclaveSigner,
+	TopPoolAuthor,
+	NodeMetadataProvider,
+> where
+	ShieldingKeyRepository: AccessKey,
+	ShieldingKeyRepository::KeyType: ShieldingCryptoDecrypt<Error = itp_sgx_crypto::Error>
+		+ ShieldingCryptoEncrypt<Error = itp_sgx_crypto::Error>,
+	StfEnclaveSigner: StfEnclaveSigning,
+	TopPoolAuthor: AuthorApi<H256, H256> + Send + Sync + 'static,
+	NodeMetadataProvider: AccessNodeMetadata,
+	NodeMetadataProvider::MetadataType: IMPCallIndexes + TeerexCallIndexes + VCMPCallIndexes,
+{
+	type Call: Decode + Encode + Clone;
+
+	fn call_index(&self, call: Self::Call) -> [u8; 2];
+
+	fn call_index_from_metadata(
+		&self,
+		metadata_type: &NodeMetadataProvider::MetadataType,
+	) -> Result<[u8; 2], MetadataError>;
+
+	fn is_target_call(&self, call: Self::Call, node_metadata: &NodeMetadataProvider) -> bool {
+		node_metadata
+			.get_from_metadata(|m| match self.call_index_from_metadata(m) {
+				Ok(call_index) => self.call_index(call) == call_index,
+				Err(_e) => false,
+			})
+			.unwrap_or(false)
+	}
+
+	fn decode(
+		&self,
+		input: &mut &[u8],
+	) -> Result<ParentchainUncheckedExtrinsic<Self::Call>, CodecError> {
+		ParentchainUncheckedExtrinsic::<Self::Call>::decode(input)
+	}
+
+	/// extrinisc in this function should execute successfully on parentchain
+	fn execute(
+		&self,
+		context: &IndirectCallsExecutor<
+			ShieldingKeyRepository,
+			StfEnclaveSigner,
+			TopPoolAuthor,
+			NodeMetadataProvider,
+		>,
+		extrinsic: UncheckedExtrinsicV4<Self::Call, SubstrateDefaultSignedExtra<PlainTip>>,
+	) -> Result<ExecutionStatus, Error>;
+
+	fn decode_and_execute(
+		&self,
+		context: &IndirectCallsExecutor<
+			ShieldingKeyRepository,
+			StfEnclaveSigner,
+			TopPoolAuthor,
+			NodeMetadataProvider,
+		>,
+		input: &mut &[u8],
+	) -> Result<ExecutionStatus, Error> {
+		if let Ok(xt) = self.decode(input) {
+			if self.is_target_call(xt.function.clone(), context.node_meta_data_provider.as_ref()) {
+				self.execute(context, xt)
+			} else {
+				Ok(ExecutionStatus::NextExecutor)
+			}
+		} else {
+			Ok(ExecutionStatus::NextExecutor)
+		}
+	}
+}
+
+pub(crate) trait DecorateExecutor<
+	ShieldingKeyRepository,
+	StfEnclaveSigner,
+	TopPoolAuthor,
+	NodeMetadataProvider,
+>
+{
```

Comment:
I think the `DecorateExecutor` is not necessary. use the `decode_and_execute` function in the trait `Executor` directly.

Reply:
After discussing with @DogLi, we decided to keep the trait `DecorateExecutor`. Since there is a associated type `Call` in trait `Executor`, `Executor` can't be directly inserted into Vector in `indriect_call_executor`, which requires `DecorateExecutor`. 
Now I just put the implementation of the function `decode_and_execute` into `DecorateExecutor`, it makes the code more reasonable.